userscript-preferences.js

```diff
@@ -0,0 +1,230 @@
+                }
+                break;
+            case "radio":
+                // let item_id = this_id + "_" + i;
+                for ( let j=0; j<pref.options.length; j++ ) {
+                    let value = pref.options[j];
+                    let checked = value === this_value ? "checked=checked" : "";
+                    dd.append(
+                        "<label>"+
+                        `<input data-userscript-pref='${this_id}' name='${this_id}' type='radio' value='${value}' ${checked} />`+
+                        value+
+                        "</label>"+
+                        "<br />"
+                    );
+                }
+                break;
+        }
+    }
+
+    let save = $( `<button data-userscript-save='${def.key}'>Save "${label}" Settings</button>` );
+    save.click( function() {
+        _save_userscript_settings(def);
+    } );
+
+    core_prefs.append(
+        "<dt></dt>",
+        $("<dd></dd>").append( save )
+    );
+}
+
+function _save_userscript_settings(def) {
+    "use strict";
+
+    let values = {};
+
+    for ( let i=0; i<def.options.length; i++ ) {
+        let pref    = def.options[i];
+        let this_id = def.key + "_" + pref.key;
+
+        switch ( pref.type ) {
+            case "text":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]`).first().val();
+                break;
+            case "checkbox":
+                let is_checked = $(`input[data-userscript-pref=${this_id}]`).first().prop("checked");
+                values[ pref.key ] = is_checked ? true : false;
+                break;
+            case "array_checkbox":
+                let all_checkboxes = $(`input[data-userscript-pref=${this_id}]`);
+                values[ pref.key ] = {};
+                all_checkboxes.map( function() {
+                    values[ pref.key ][ $(this).prop("name") ] = $(this).prop("checked") ? true : false;
+                } );
+                break;
+            case "radio":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]:checked`).val();
+                break;
+
+        }
+    }
+
+    localStorage.setItem(
+        def.key,
+        JSON.stringify( values )
+    );
+
+    if ( def.hasOwnProperty("callback") ) {
+        def["callback"]( values );
+    }
+
+    $( `button[data-userscript-save=${def.key}]` ).first().append(
+        "<span class='fa fa-check-circle' style='color: green; padding-left: 0.5em'></span>"
+    );
+
+    window.setTimeout(
+        function() {
+            $( `button[data-userscript-save=${def.key}]` )
+                .first()
+                .find( "span[class^=fa]" )
+                .remove();
+        },
+        1000
+    );
+}
+
+var player_name;
+
+function get_player_storage_prefix(script_prefix) {
+    // Get the player's name, to let us store different session data for different player characters.
+    if (! player_name) {
+        player_name = $("#player-name").text();
```

Comment:
Heh. _Technically_, Tau frowns upon an individual having two player accounts, so this may not be needed. :-)

(I originally added this to taustation-tools-common.js as a "just in case", to preemptively avoid hypothetical problems. Lately, I've thought about removing it for the ToS reason; however, I decided to leave it be for now, so I didn't have issues moving old-style localStorage config keys to ones lacking the player name.  Since this options code is new, though, "merge existing config" isn't an issue, so we could omit the player name from the start.)